File: 02-arrays/easy/11-longestSubarraySum.cpp
#include <bits/stdc++.h>
using namespace std;

// // Problem Statement : Given an array and a sum k, we need to print the
// //  length of the longest subarray that sums to k.
int getLongestSubarray(vector<int> &arr, long long k) {
    int n = arr.size();
    int len = 0;
    for (int i = 0; i < n; i++) {
        long long sum = 0;
        for (int j = i; j < n; j++) {
            sum += arr[j];
            if (sum == k)
                len = max(len, j - i + 1);
        }
    }
    return len;
}

int getLongestSubarray_PrifixSum(vector<int> &arr, long long k) {
    int n = arr.size();
    int maxLen = 0;
    long long sum = 0;
    map<long long, int> prefixSum;

    for (int i = 0; i < n; i++) {
        // calculate the prefix sum till index i:
        sum += arr[i];
        if (sum == k)
            maxLen = max(maxLen, i + 1);
        long long rem = sum - k;
        if (prefixSum.find(rem) != prefixSum.end())
            maxLen = max(maxLen, i - prefixSum[rem]);
        if (prefixSum.find(sum) == prefixSum.end())
            prefixSum[sum] = i;
    }
    return maxLen;
}

int getLongestSubarray_LRPointer(vector<int> &arr, long long k) {
    int n = arr.size();
    int left = 0;
    int right = 0;
    int maxLen = 0;
    long long sum = arr[0];

    while (right < n) {
        while (sum > k && left <= right) {
            sum = sum - arr[left];
            left++;
        }
        if (sum == k)
            maxLen = max(maxLen, right - left + 1);

        right++;
        if (right < n)
            sum += arr[right];
    }
    return maxLen;
}

File: 02-arrays/easy/11-longestSubarraySum.go
package main

import (
	"fmt"
)

func LongestSubarraySum(arr []int, k int) int {
	n := len(arr)
	left := 0
	right := 0
	sum := arr[0]
	max := 0
	for right < n {
		for sum > k && left <= right {
			sum -= arr[left]
			left++
		}

		if sum == k {
			if max < right-left+1 {
				max = right - left + 1
			}
		}
		right++
		if right < n {
			sum += arr[right]
		}
	}
	return max
}

func LongestSubarraySumPrefix(arr []int, k int) int {
	n := len(arr)
	var sum int64 = 0
	max := 0
	prefixSum := make(map[int64]int)

	for i := 0; i < n; i++ {
		sum += int64(arr[i])
		if sum == int64(k) && max < i+1 {
			max = i + 1
		}
		var rem int64 = sum - int64(k)
		if index, ok := prefixSum[rem]; ok {
			if max < i-index {
				max = i - index
			}
		}
		if _, ok := prefixSum[sum]; !ok {
			prefixSum[sum] = i
		}
	}
	return max
}

func main() {
	arr := []int{10, 5, 2, 7, 1, 9}
	res := LongestSubarraySumPrefix(arr, 15)
	fmt.Println(res)
}
